op-batcher: lock state mutex while computing sync actions

geoknee opened the pull request:
Closes #13332 and Closes #13280

geoknee committed:
lock state mutex while  computing sync actions

sebastianst reviewed (changes requested):
We indeed have inconsistent concurrent access patterns to the `channelManager`, but I think we need to fix this more fundamentally.

E.g. in lines 456-457 in `driver.go`, we call
```
				l.state.pruneSafeBlocks(syncActions.blocksToPrune)
				l.state.pruneChannels(syncActions.channelsToPrune)
```
while we don't lock the state mutex. Those methods don't internally lock it. Plus, there's quite a few actions on the `state` that we want to perform atomically, E.g. calculate the sync actions and apply them. This is currently not possible. It's also not a good pattern that the `channelManager` itself holds its `sync.Mutex` while there can be multiple operations on it that should happen atomically.

My proposal to fix this properly is to move the `channelManager` mutex out as a `stateMu sync.Mutex` variable (or we take the opportunity to rename the `state` to `channelMgr` and the mutex then to `channelMu` or so) into the driver and wherever we need concurrent access to the `channelManager`, the mutex is locked and `defer l.channelMu.Unlock()`'d from the driver.

Then in the main loop, we can lock the mutex in line 434, as is happening in this PR, but only unlock in line 459, right before loading blocks. This way, the whole sync computation and pruning of blocks and channels is atomic. Feels like this could then also be moved into its own function.

If we remove the mutex, we need to check that all accesses that are currently secured by the mutex stay concurrency-safe. This is currently the case with `Clear`, `TxFailed`, `TxConfirmed`, `TxData`, and `AddL2Block`.

An alternative to moving the mutex out of the `channelManager` is to create a method on it that encapsulates all the operations that are currently happening in the driver's main loop between lines 434-459. But not sure if this is prettier or even possible.

samlaf commented on op-batcher/batcher/driver.go:
```diff
@@ -430,8 +430,14 @@ func (l *BatchSubmitter) mainLoop(ctx context.Context, receiptsCh chan txmgr.TxR
 				continue
 			}
 
+			// Lock the state to prevent any changes while we're computing sync actions
+			l.state.mu.Lock()
+			blocks := l.state.blocks
+			channels := l.state.channelQueue
+			l.state.mu.Unlock()
```
Don't think this solves the race condition, since the  computeSyncActions function still calls functions on the channels. Here's what I had done: https://github.com/ethereum-optimism/optimism/pull/13169/files#diff-2a830eb26f5b3a174fa39ce7e267cbabe362ccb08b354c1b4e1abb9f60283808R115

samlaf replied:
You might actually just want to make the `computeSyncActions` function a method on the channelManager?

geoknee replied:
Keen to keep this function pure to ensure it remains easily unit testable.

sebastianst replied:
we've changed the locking to happen around the sync computation and pruning

geoknee committed:
rename batcher.state to batcher.channelManager

Closes https://github.com/ethereum-optimism/optimism/issues/13280

geoknee committed:
move channel manager mutex up into driver

encapsulate imperative code and mutex locking/unlocking into new method executeSyncActions

geoknee committed:
don't use pointer to mutex

geoknee committed:
rename

geoknee commented:
@sebastianst @samlaf I updated the PR along @sebastianst 's suggestion. 

The new mutex basically ensures the `mainLoop` and `receiptsLoop` get exclusive access to the `channelManager`  for the duration of one iteration of their event loop.

sebastianst commented on op-batcher/batcher/driver.go:
```diff
@@ -114,8 +114,9 @@ type BatchSubmitter struct {
 
-	state         *channelManager
-	prevCurrentL1 eth.L1BlockRef // cached CurrentL1 from the last syncStatus
+	channelMgrMutex sync.Mutex // guards channelMgr and prevCurrentL1
```
a conventional shorter name would be 
```suggestion
	channelMgrMu sync.Mutex // guards channelMgr and prevCurrentL1
```
up to you

geoknee replied:
I think I'll leave this as is for now so it matches other mutexes in the same struct.

sebastianst commented on op-batcher/batcher/driver.go:
```diff
@@ -387,6 +388,47 @@ func (l *BatchSubmitter) setTxPoolState(txPoolState TxPoolState, txPoolBlockedBl
+		l.channelMgr.pruneChannels(syncActions.channelsToPrune)
+	}
+
+	if syncActions.blocksToLoad != nil {
+		// Get fresh unsafe blocks
+		if err := l.loadBlocksIntoState(l.shutdownCtx, syncActions.blocksToLoad.start, syncActions.blocksToLoad.end); errors.Is(err, ErrReorg) {
+			l.Log.Warn("error loading blocks, clearing state and waiting for node sync", "err", err)
+			l.waitNodeSyncAndClearState()
+			return
+		}
+	}
```
We're now holding the mutex while loading blocks, which can take time, especially at startup. This is a significant change in locking behavior. I think it's fine, but we should have an eye on it.

The reorg case might be most concerning, since the `waitNodeSyncAndClearState` may take some time.

geoknee replied:
Yes, a valid concern. I think we have to prefer atomicity to "performance" here. The main loop is definitely the most important now, since the receipts handling loop only deals with quite edge cases which the main loop would eventually catch. The throttling loop could be starved for longer than before, though, which could be a concern.

geoknee replied:
In the end the flakiness of the test with this "long lock holding" approach was a signal that we aren't ready to make that change. So I have reverted to a pattern which releases the lock more often, between pruning and loading blocks (and in fact between each block that is loaded).

sebastianst commented on op-batcher/batcher/driver.go:
```diff
@@ -387,6 +388,47 @@ func (l *BatchSubmitter) setTxPoolState(txPoolState TxPoolState, txPoolBlockedBl
+		}
+	}
+
+	l.publishStateToL1(queue, receiptsCh, daGroup, l.Config.PollInterval)
```
This is non-blocking, always publishing async via the queue, right?

geoknee replied:
Yes. We could release the mutex before calling this, though.

sebastianst commented on op-batcher/batcher/driver.go:
```diff
@@ -647,7 +660,7 @@ func (l *BatchSubmitter) clearState(ctx context.Context) {
 			return false
 		} else {
 			l.Log.Info("Clearing state with safe L1 origin", "origin", l1SafeOrigin)
-			l.state.Clear(l1SafeOrigin)
+			l.channelMgr.Clear(l1SafeOrigin)
```
don't we need to guard this call with the mutex?

geoknee replied:
This function is only called:
* by StartBatchSubmitting _before_ any goroutines are launched and 
* by syncAndPublish, which is already holding the lock

So we shouldn't add a guard here, because it is unecessary and would cause a deadlock.

sebastianst commented on op-batcher/batcher/driver.go:
```diff
@@ -668,7 +681,7 @@ func (l *BatchSubmitter) clearState(ctx context.Context) {
 			}
 		case <-ctx.Done():
 			l.Log.Warn("Clearing state cancelled")
-			l.state.Clear(eth.BlockID{})
+			l.channelMgr.Clear(eth.BlockID{})
```
don't we need to guard this call with the mutex?

geoknee replied:
Same as above, this function is only called:

* by StartBatchSubmitting before any goroutines are launched and
* by syncAndPublish, which is already holding the lock

So we shouldn't add a guard here, because it is unecessary and would cause a deadlock.

sebastianst commented on op-batcher/batcher/driver.go:
```diff
@@ -686,7 +699,7 @@ func (l *BatchSubmitter) publishTxToL1(ctx context.Context, queue *txmgr.Queue[t
 
 	// Collect next transaction data. This pulls data out of the channel, so we need to make sure
 	// to put it back if ever da or txmgr requests fail, by calling l.recordFailedDARequest/recordFailedTx.
-	txdata, err := l.state.TxData(l1tip.ID())
+	txdata, err := l.channelMgr.TxData(l1tip.ID())
```
don't we need to guard this call with the mutex?

geoknee replied:
In the current pattern, this is only called inside `syncAndPublish` which already holds the lock.

sebastianst commented on op-batcher/batcher/driver.go:
```diff
@@ -870,18 +883,22 @@ func (l *BatchSubmitter) recordFailedDARequest(id txID, err error) {
 	if err != nil {
 		l.Log.Warn("DA request failed", logFields(id, err)...)
 	}
-	l.state.TxFailed(id)
+	l.channelMgr.TxFailed(id)
```
don't we need to guard this call with the mutex?

geoknee replied:
This is called inside `syncAndPublish`, which already holds the lock.

sebastianst commented on op-batcher/batcher/driver.go:
```diff
@@ -870,18 +883,22 @@ func (l *BatchSubmitter) recordFailedDARequest(id txID, err error) {
 func (l *BatchSubmitter) recordFailedTx(id txID, err error) {
+	l.channelMgrMutex.Lock()
+	defer l.channelMgrMutex.Unlock()
 	l.Log.Warn("Transaction failed to send", logFields(id, err)...)
-	l.state.TxFailed(id)
+	l.channelMgr.TxFailed(id)
 }
 
 func (l *BatchSubmitter) recordConfirmedTx(id txID, receipt *types.Receipt) {
+	l.channelMgrMutex.Lock()
+	defer l.channelMgrMutex.Unlock()
 	l.Log.Info("Transaction confirmed", logFields(id, receipt)...)
 	l1block := eth.ReceiptBlockID(receipt)
-	l.state.TxConfirmed(id, l1block)
+	l.channelMgr.TxConfirmed(id, l1block)
```
here you add the guards

geoknee replied:
Yes there are other places we could place them, but I'm trying to stick to the `defer` pattern and to place the guards near to the explicit access of the `channelMgr`.

sebastianst commented on op-batcher/batcher/driver.go:
```diff
@@ -514,7 +527,7 @@ func (l *BatchSubmitter) throttlingLoop(ctx context.Context) {
 			l.Log.Error("Can't reach sequencer execution RPC", "err", err)
 			return
 		}
-		pendingBytes := l.state.PendingDABytes()
+		pendingBytes := l.channelMgr.PendingDABytes()
```
It may be fine not to guard this call. But strictly speaking we'd need to... at least an `RLock()`.

geoknee replied:
If it's the _only_ instance of a read, it might not be worth making it a read only lock.

samlaf commented on op-batcher/batcher/driver.go:
```diff
@@ -114,8 +114,9 @@ type BatchSubmitter struct {
-	state         *channelManager
-	prevCurrentL1 eth.L1BlockRef // cached CurrentL1 from the last syncStatus
+	channelMgrMutex sync.Mutex // guards channelMgr and prevCurrentL1
+	channelMgr      *channelManager
```
yay! finally :)

geoknee commented:
There are different approaches to optimising when the mutex is held. For now we are minded to go for a simple route, where the main loop holds the lock for a long time, even doing network calls and waiting for l1 transactions. The receipts loop would then be blocked, but it is not a important as the main loop anyway. 

We can always return and optimise things to be more granular. But to be safe it makes sense to make each goroutines actions atomic, therefore holding the lock and excluding the other goroutines.

geoknee committed:
throttlingLoop uses channelMgrMutex

geoknee committed:
unblock throttling loop by sending pendingBytes on a channel

geoknee committed:
disable proposer in 4844 system test

it mostly adds noise to the logs

geoknee committed:
remove unused ticker

geoknee committed:
Revert "remove unused ticker"

This reverts commit b2e9762c5e6e92e8b62771439d864d9985dcc943.

geoknee committed:
reinstate throttle on tick

used cached value

geoknee committed:
make the main loop release the chMgr lock more often

geoknee committed:
simplify

geoknee committed:
add some more mutex calls

geoknee committed:
simplify some more

geoknee committed:
push mutex locking down into l.publishTxToL1

geoknee commented on op-batcher/batcher/channel_manager.go:
```diff
@@ -29,7 +28,6 @@ type ChannelOutFactory func(cfg ChannelConfig, rollupCfg *rollup.Config) (derive
 // channel.
 // Public functions on channelManager are safe for concurrent access.
 type channelManager struct {
-	mu          sync.Mutex
```
This mutex has been removed in favor of one in the parent struct (the driver), so that it can have greater control over which operations to drive atomically.

geoknee commented on op-batcher/batcher/driver.go:
```diff
@@ -105,7 +105,7 @@ type BatchSubmitter struct {
 	cancelKillCtx     context.CancelFunc
 
-	l2BlockAdded chan struct{} // notifies the throttling loop whenever an l2 block is added
+	pendingBytesUpdated chan int64 // notifies the throttling with the new pending bytes
```
This was upgraded from sending empty signal to sending the actual information the throttling loop needs to do its job. This allows the throttling loop to not require the channel manager mutex, since it won't read from there anymore.

geoknee commented on op-batcher/batcher/driver.go:
```diff
@@ -550,12 +564,14 @@ func (l *BatchSubmitter) throttlingLoop(ctx context.Context) {
 		}
 	}
 
+	cachedPendingBytes := int64(0)
```
We use this cache in the case where the ticker fires, we just use the previous value. This allows us to throttle a new sequencer which we didn't talk to before.

geoknee committed:
do not signal the throttling loop inside publishTxToL1

This is a change in behaviour that is outside the scope of this PR.

sebastianst commented on op-batcher/batcher/driver.go:
```diff
@@ -681,14 +688,17 @@ func (l *BatchSubmitter) clearState(ctx context.Context) {
 			l.Log.Warn("Clearing state cancelled")
+			l.channelMgrMutex.Lock()
 			l.channelMgr.Clear(eth.BlockID{})
+			l.channelMgrMutex.Unlock()
```
```suggestion
			l.channelMgrMutex.Lock()
			defer l.channelMgrMutex.Unlock()
			l.channelMgr.Clear(eth.BlockID{})
```
you can do this since you return right after

geoknee replied:
[2bcc0ae](https://github.com/ethereum-optimism/optimism/pull/13330/commits/2bcc0ae86aadc9029c099973bb55c149efe1f189)

sebastianst commented on op-batcher/batcher/driver.go:
```diff
@@ -660,7 +665,9 @@ func (l *BatchSubmitter) clearState(ctx context.Context) {
 			l.Log.Info("Clearing state with safe L1 origin", "origin", l1SafeOrigin)
+			l.channelMgrMutex.Lock()
 			l.channelMgr.Clear(l1SafeOrigin)
+			l.channelMgrMutex.Unlock()
```
```suggestion
			l.channelMgrMutex.Lock()
			defer l.channelMgrMutex.Unlock()
			l.channelMgr.Clear(l1SafeOrigin)
```

geoknee replied:
[2bcc0ae](https://github.com/ethereum-optimism/optimism/pull/13330/commits/2bcc0ae86aadc9029c099973bb55c149efe1f189)

sebastianst commented on op-batcher/batcher/driver.go:
```diff
@@ -699,7 +709,9 @@ func (l *BatchSubmitter) publishTxToL1(ctx context.Context, queue *txmgr.Queue[t
 	// to put it back if ever da or txmgr requests fail, by calling l.recordFailedDARequest/recordFailedTx.
+	l.channelMgrMutex.Lock()
 	txdata, err := l.channelMgr.TxData(l1tip.ID())
+	l.channelMgrMutex.Unlock()
```
maybe it's worth moving this into a `func() {...}()` call so we can use `defer`

geoknee replied:
I don't think that is worth it, here: we would need to declare txdata outside the anonymous fn and generally it's more boilerplate for no change in functionality.

sebastianst replied:
It would unlock the mutex in case of a panic, which is slightly better. But then, if we have a panic inside `TxData`, the batcher is broken. So it's fine for now. Just generally feeling better about always unlocking in a `defer`.

sebastianst reviewed (commented):
Looking good now, just minor nits about mutex usage.

geoknee committed:
prefer deferring mutex unlock
